### PR TITLE
Sparse hessian

### PR DESCRIPTION
**Pull request by amontoison**
@tmigot I have the wrong result when I use `autoauto_color_hessian!`.
I suspect an error in SparseDiffTools.jl. :thinking:

### CODE REVIEW

**Comment by dpo** [comment]
Could we see a benchmark with JuMP just to know where this stands?

**Comment by amontoison** [comment]
@tmigot Is it possible to do have a benchmark with the `SparseHessian` backend based on Symbolics.jl for Dominique?

**Comment by tmigot** [comment]
The benchmark JuMP vs Symbolics Hessian:
https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/blob/benchmark/benchmark/2023-04-09_adnlpmodels_benchmark_hess_nscal_100_mono.png
for problems with about 100 variables and constraints.

**Review by tmigot — CHANGES_REQUESTED** [review]
Can you also change these two lines:
https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/blob/327bd6373be23901fbac2e87ec7cf142326699f0/src/ad.jl#L77

https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/blob/327bd6373be23901fbac2e87ec7cf142326699f0/src/ad.jl#L108

to have this new backend by default?

**Comment by tmigot** [comment]
2023-04-27_adnlpmodels_benchmark_hess_nscal_1000_mono
Not bad!

File: src/sparse_hessian.jl [hunk 1]
```diff
@@ -1,3 +1,100 @@
+## ----- SparseDiffTools -----
```
Comment (tmigot) [inline src/sparse_hessian.jl:1]: Should we remove this comment?

Comment (amontoison) [reply src/sparse_hessian.jl:1]: Yes, I also propose the rename this backend `SparseADHessian`.

File: src/ad.jl [hunk 2]
```diff
@@ -74,7 +74,7 @@ function ADModelBackend(
   jprod_backend::Type{JvB} = ForwardDiffADJprod,
   jtprod_backend::Type{JtvB} = ForwardDiffADJtprod,
   jacobian_backend::Type{JB} = SparseForwardADJacobian,
-  hessian_backend::Type{HB} = ForwardDiffADHessian, # sparse alternative: SparseSymbolicsADHessian
+  hessian_backend::Type{HB} = ForwardDiffADHessian, # sparse alternative: SparseADHessian
```
Comment (tmigot) [inline src/ad.jl:77]:
```suggestion
  hessian_backend::Type{HB} = SparseADHessian,
```

File: src/ad.jl [hunk 3]
```diff
@@ -105,7 +105,7 @@ function ADModelNLSBackend(
...
```
Comment (tmigot) [inline src/ad.jl:108]:
```suggestion
  hessian_backend::Type{HB} = SparseADHessian,
```

**Review by tmigot — CHANGES_REQUESTED** [review]

**Comment by amontoison** [comment]
> Not bad! Sparse is this one, Forward is the dense ForwardDiff.
> 
> 2023-04-27_adnlpmodels_benchmark_hess_nscal_1000_mono

Nice! Can you add the `Symbolic` version too?
We didn't optimize the directional derivative of the gradient this afternoon so it's already great!

File: src/sparse_hessian.jl [hunk 4]
```diff
@@ -1,3 +1,98 @@
...
+  b::SparseADHessian,
+  nlp::ADModel,
+  x::AbstractVector,
+  obj_weight::Real,
+  vals::AbstractVector,
+)
+  ℓ(x) = obj_weight * nlp.f(x)
+  sparse_hess_coord!(ℓ, b, x, vals)
+end
+
```
Comment (tmigot) [inline src/sparse_hessian.jl:95]: The tests should pass after this:
```suggestion

function hess_coord!(
  b::SparseADHessian,
  nlp::ADModel,
  x::AbstractVector,
  j::Integer,
  vals::AbstractVector{T},
) where {T}
  y = zeros(T, nlp.meta.nnln)
  for (w, k) in enumerate(nlp.meta.nln)
    y[w] = k == j ? 1 : 0
  end
  obj_weight = zero(T)
  ℓ = get_lag(nlp, b, obj_weight, y)
  sparse_hess_coord!(ℓ, b, x, vals)
  return vals
end
```

File: src/sparse_hessian.jl [hunk 5]
```diff
@@ -1,3 +1,98 @@
...
+function hess_coord!(
+  b::SparseADHessian,
+  nlp::ADModel,
+  x::AbstractVector,
+  obj_weight::Real,
+  vals::AbstractVector,
+)
+  ℓ(x) = obj_weight * nlp.f(x)
```
Comment (tmigot) [inline src/sparse_hessian.jl:92]:
```suggestion
  ℓ = get_lag(nlp, b, obj_weight)
```
as `nlp.f` is not defined for NLS

**Review by tmigot — APPROVED** [review]
Looks good to me !
Next steps:
- [ ] improve hprod;
- [ ] add our own algo for thé coloration.

**Comment by jbcaillau** [comment]
@amontoison @tmigot Hi guys, looks good: any release available for testing on sparse Jacobians / Hessians?

**Comment by amontoison** [comment]
Hi @jbcaillau!
Tangi, do you think that we can do a release after #139?

**Comment by tmigot** [comment]
Actually, we just need to try and solve https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/issues/140 first, but next release should be in a week or 2. I will keep you guys updated when it is done.

**Comment by jbcaillau** [comment]
@tmigot @amontoison hi there, any news on this? Actually, can you please update me on what might be pulled to test solving with sparse features?

**Comment by tmigot** [comment]
Hi @jbcaillau ! ADNLPModels 0.6.2 has the sparse Hessian and Jacobian implemented. By the end of the month, there will be another upgraded version with improved `jprod`, `jtprod` and `hprod` and faster `Hessian/jacobian`.

**Comment by jbcaillau** [comment]
Thanks @tmigot for the confirmation. I am forwarding to the ct team that they can rely on 0.6.2 to benchmark sparse derivatives 👍🏽

**Comment by amontoison** [comment]
@jbcaillau 
We released [JuliaHSL](https://licences.stfc.ac.uk/product/julia-hsl) this week.
If you use Ipopt.jl, you will probably see a huge difference.

**Comment by jbcaillau** [comment]
thanks @amontoison : does it require to apply for an academic license?

**Comment by amontoison** [comment]
Yes, you need to apply for an academic license.

**Comment by jbcaillau** [comment]
✅ order passed 🤞🏾

**Comment by jbcaillau** [comment]
Hi @tmigot @amontoison! Not sure to be on the right issue, but anyway: as JuliaCon approaches, we are on the way to announce the last release of [control-toolbox](https://control-toolbox.org). Any news of performance increase for sparse Hessians?

**Comment by tmigot** [comment]
Hi @jbcaillau ! I will make a last attempt on Monday to extend to the Lagrangian what we did there https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/pull/171 . I keep you updated.

**Pull request merged**